### PR TITLE
Change db null generation

### PR DESCRIPTION
**Pull request by darkruby501**
┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1201302964208280/1205763261340741) by [Unito](https://www.unito.io)

### CODE REVIEW

File: packages/lesswrong/lib/collections/revisions/collection.ts [hunk 1]
```diff
@@ -48,6 +48,11 @@ Revisions.checkAccess = async (user: DbUser|null, revision: DbRevision, context:
   const { major: majorVersion } = extractVersionsFromSemver(revision.version)
   const collection = getCollection(collectionName);
   const documentId = revision.documentId;
+
+  if (!documentId) {
+    return false
```
Comment (b0b3rt) [inline packages/lesswrong/lib/collections/revisions/collection.ts:53]: Annoyingly, we do actually have some revisions with missing `documentIds`.  If they don't belong to the current user attempting to access them (by their `userId`), then I don't know of a safer thing to do than to deny them permission to access the revision.  (I'm not actually sure what happens if you pass `null` into either a `loader.load` or a `collection.findOne` - with mongo I'm pretty sure the `findOne` would've returned a totally random document, which would be bad, but I doubt that's happening with postgres.)

File: packages/lesswrong/lib/collections/revisions/schema.ts [hunk 2]
```diff
@@ -154,7 +154,7 @@ const schema: SchemaType<DbRevision> = {
         // suggestion. Original contents is only visible to people who are invited 
         // to collaborative editing. (This is only relevant for posts, but supporting
         // it means we need originalContents to default to unviewable)
-        if (document.collectionName === "Posts") {
+        if (document.collectionName === "Posts" && document.documentId) {
```
Comment (b0b3rt) [inline packages/lesswrong/lib/collections/revisions/schema.ts:157]: Actually, looking at the context here, I think this should be treated more like the `Revisions.checkAccess` logic.  Need to come back to this one and think about it a bit more.

File: packages/lesswrong/lib/collections/users/helpers.tsx [hunk 3]
```diff
@@ -88,7 +88,7 @@ export const userIsSharedOn = (currentUser: DbUser|UsersMinimumInfo|null, docume
     return (
       document.sharingSettings?.anyoneWithLinkCan
       && document.sharingSettings.anyoneWithLinkCan !== "none"
-      && _.contains((document as DbPost).linkSharingKeyUsedBy, currentUser._id)
+      && ((document as DbPost).linkSharingKeyUsedBy)?.includes(currentUser._id) //TODO: Robert, check this one especially
```
Comment (b0b3rt) [inline packages/lesswrong/lib/collections/users/helpers.tsx:91]: Yep, seems like the correct translation

File: packages/lesswrong/lib/index.ts [hunk 4]
```diff
@@ -4,9 +4,14 @@ import './vulcan-users/index';
...
 // schema utils
 import './utils/schemaUtils'
+
+// collection utils
+import './collectionUtils'
+
 // Permissions
 import './permissions';
 
+
```
Comment (b0b3rt) [inline packages/lesswrong/lib/index.ts:14]: Undo the changes to this file now that we've moved the SimpleSchema.extendOptions elsewhere

File: packages/lesswrong/lib/utils/schemaUtils.ts [hunk 5]
```diff
@@ -394,3 +394,35 @@ export function googleLocationToMongoLocation(gmaps: AnyBecauseTodo) {
...
+      throw new Error(`Error updating: ${fieldName} cannot be null or missing`);
+    }
+  };
+
+  return {
+    defaultValue: defaultValue,
+    onCreate: fillIfMissing,
+    onUpdate: throwIfSetToNull,
+    canAutofillDefault: true,
+    nullable: false
```
Comment (b0b3rt) [inline packages/lesswrong/lib/utils/schemaUtils.ts:426]: Note for other reviewers: we now set `nullable: false` by default when using `schemaDefaultValue`.  This can still be overriden by explicitly setting `nullable: true` after spreading the `schemaDefaultValue` output in a field definition, if you want to explicitly allow `null` values for that column.

File: packages/lesswrong/lib/utils/schemaUtils.ts [hunk 6]
```diff
@@ -394,3 +394,35 @@ export function googleLocationToMongoLocation(gmaps: AnyBecauseTodo) {
     coordinates: [gmaps.geometry.location.lng, gmaps.geometry.location.lat]
   }
 }
+export function schemaDefaultValue<T extends DbObject>(defaultValue: any): Partial<CollectionFieldSpecification<T>> {
```
Comment (b0b3rt) [inline packages/lesswrong/lib/utils/schemaUtils.ts:397]: This was moved to break an import cycle.

File: packages/lesswrong/lib/vulcan-lib/utils.ts [hunk 7]
```diff
@@ -158,7 +158,8 @@ export const slugify = function (s: string): string {
   return slug;
 };
 
-export const getDomain = function(url: string): string|null {
+export const getDomain = function(url: string | null ): string|null {
```
Comment (b0b3rt) [inline packages/lesswrong/lib/vulcan-lib/utils.ts:161]: nit: extra whitespace after `| null` in the parameter type

File: packages/lesswrong/server/apolloServer.ts [hunk 8]
```diff
@@ -201,12 +201,12 @@ export function startWebserver() {
     }
     
     const currentUser = await getUserFromReq(req)
-    if (!currentUser || !getUserEmail(currentUser)){
+    if (!currentUser || !currentUser.email){
```
Comment (b0b3rt) [inline packages/lesswrong/server/apolloServer.ts:204]: This... does not seem like a safe substitution?

File: packages/lesswrong/server/apolloServer.ts [hunk 9]
```diff
@@ -201,12 +201,12 @@ export function startWebserver() {
...
       res.status(403).send("Not logged in or current user has no email address")
       return
     }
     
-    const eagApp = await getEAGApplicationData(currentUser.email)
+    const eagApp = await getEAGApplicationData(currentUser.email)
```
Comment (b0b3rt) [inline packages/lesswrong/server/apolloServer.ts:209]: What's the diff here?

File: packages/lesswrong/server/apolloServer.ts (hunk 8 above)
Comment (darkruby501) [reply packages/lesswrong/server/apolloServer.ts:204]: I thought I put that back as it was

**Comment by jimrandomh** [comment]
The LW dev DB is currently borked. I don't know whether this PR is related, but this PR is pretty high risk and we definitely don't want to merge it without seeing it migrate the dev DB successfully and end in an unborked state..

**Comment by b0b3rt** [comment]
TODO: fix all the `upsert` and `ON CONFLICT` queries, so that they don't `coalesce` on columns that are no longer null.

File: packages/lesswrong/server/editor/make_editable_callbacks.ts [hunk 10]
```diff
@@ -23,6 +23,7 @@ import {notifyUsersAboutMentions, PingbackDocumentPartial} from './mentions-noti
 import {isBeingUndrafted, MaybeDrafteable} from './utils'
 import { Comments } from '../../lib/collections/comments'
 import { cheerioParse } from '../utils/htmlUtil'
+import { or } from 'fp-ts/lib/Predicate'
```
Comment (oetherington) [inline packages/lesswrong/server/editor/make_editable_callbacks.ts:26]: Unused import

File: packages/lesswrong/server/callbacks.ts [hunk 11]
```diff
@@ -65,7 +65,7 @@ getCollectionHooks("Users").updateAsync.add(function userEditDeleteContentCallba
 });
 
 getCollectionHooks("Users").editAsync.add(function userEditBannedCallbacksAsync(user: DbUser, oldUser: DbUser) {
-  if (new Date(user.banned) > new Date() && !(new Date(oldUser.banned) > new Date())) {
+  if (user.banned && oldUser.banned && new Date(user.banned) > new Date() && !(new Date(oldUser.banned) > new Date())) {
```
Comment (oetherington) [inline packages/lesswrong/server/callbacks.ts:68]: This is probably wrong. I think we're trying to detect the case where `user` is banned but `oldUser` is not (`new Date(null)` is the Unix epoch and `new Date(undefined)` is "Invalid date", so `!(new Date(oldUser.banned) > new Date())` is true in either case).

File: packages/lesswrong/lib/sql/InsertQuery.ts [hunk 12]
```diff
@@ -138,6 +138,11 @@ class InsertQuery<T extends DbObject> extends Query<T> {
         } else if (key === "createdAt" && !(item as AnyBecauseTodo)[key]) {
           (item as AnyBecauseTodo)[key] = new Date();
         }
+        const type = fields[key];
+        const value = (item as AnyBecauseTodo)[key];
+        // if ((value === null && type instanceof DefaultValueType && type.isNotNull() && type.getDefaultValueString())) {
+        //   console.log({ key, type: fields[key], arg: this.createArg((item as AnyBecauseTodo)[key] ?? null, fields[key]), value: (item as any)[key] ?? null })
+        // }
```
Comment (oetherington) [inline packages/lesswrong/lib/sql/InsertQuery.ts:145]: Just debug code?

File: packages/lesswrong/lib/sql/Query.ts [hunk 13]
```diff
@@ -19,6 +19,13 @@ class Arg {
       } else {
         this.typehint = "::JSONB[]";
       }
+    } else if (value === null && type instanceof DefaultValueType && type.isNotNull() && type.getDefaultValueString()) {
+      // console.log({type})
```
Comment (oetherington) [inline packages/lesswrong/lib/sql/Query.ts:23]: Probably remove this console.log (unless it's still needed for upsert debugging)

File: packages/lesswrong/lib/collections/tagRels/collection.ts [hunk 14]
```diff
@@ -42,14 +42,14 @@ const schema: SchemaType<DbTagRel> = {
   },
   // The user who first tagged the post with this tag
   userId: {
+    nullable: false,
```
[the remark this answers is not in the document]
Comment (darkruby501) [reply packages/lesswrong/lib/collections/tagRels/collection.ts:45]: Seems like a good change to make but probably in a different PR

**Comment by darkruby501** [comment]
**TO-DO LIST**

- Figure out where we need to do anything with `afKarma` (in Hex dashboard?)
- update SQL commands in migration from latest Hex
- merge in master to PR again
- merge in parent PR to this PR
- figure out how to do small number of indexes for ON CONFLICT constraint
- run in dev
